Sources/HAP/Server/Device.swift

```diff
@@ -145,15 +200,133 @@ public class Device {
         // other Accessory object can have an instance ID of "1" within the
         // Accessory Server.
         //
-        // TODO: SwiftFoundation in Swift 4.0 cannot encode/decode UInt64,
-        // which is the data-type we wanted to use here. We can change it back
-        // to UInt64 once the following commit has made it into a release:
-        // https://github.com/apple/swift-corelibs-foundation/commit/64b67c91479390776c43a96bd31e4e85f106d5e1
-        var idGenerator = (1...InstanceID.max).makeIterator()
-        for accessory in accessories {
+
+        // Obtain new aid's for any accessories which don't already have one
+        self.accessories = [Accessory]()
+
+        // The first accessory must be aid 1
+        accessories[0].aid = 1
+
+        addAccessories(accessories)
+    }
+
+    public func canAddAccessory(accessory: Accessory) -> Bool {
+        // HAP Spec 2.5.3.2 - Maximum 100 accessories
+        if !isBridge || accessories.count == 100 {
+            return false
+        }
+        let serialNumber = accessory.uniqueSerialNumber
+        return isUniqueSerialNumber(serialNumber, ignoring: accessory)
+    }
+
+    /// Add an array of accessories to this bridge device.
+    ///
+    /// It is an error to try and add accessories with duplicate serial numbers.
+    /// It is an error to try and add accessories to a non-bridge device.
+    /// It is an error to try and increase the number of accessories above 99.
+    public func addAccessories(_ newAccessories: [Accessory]) {
+        let totalNumberOfAccessories = accessories.count + newAccessories.count
+        precondition(
+            (isBridge && totalNumberOfAccessories <= 100) ||
+            (!isBridge && totalNumberOfAccessories == 1),
+                     "A maximum of 99 accessories can be added to a bridge")
```

Comment:
Should we have the following two preconditions instead -- we don't want to add accessories to a non-bridge device:
```swift
precondition(isBridge)
precondition(totalNumberOfAccessories <= 100)
```

Reply:
In Device.init(), all accessories, including the primary accessory are added with this function, therefore it needs to handle the case of a single non-bridge device being added at initialization. This precondition was the reason the unit tests failed.